src/librustc_mir/dataflow/impls/borrows.rs

```diff
@@ -271,26 +376,243 @@ impl<'a, 'gcx, 'tcx> BitDenotation for Borrows<'a, 'gcx, 'tcx> {
+                // When we return from the function, then all `ReScope`-style regions
+                // are guaranteed to have ended.
+                // Normally, there would be `EndRegion` statements that come before,
+                // and hence most of these loans will already be dead -- but, in some cases
+                // like unwind paths, we do not always emit `EndRegion` statements, so we
+                // add some kills here as a "backup" and to avoid spurious error messages.
+                for (borrow_index, borrow_data) in self.borrows.iter_enumerated() {
+                    if let ReScope(..) = borrow_data.region {
+                        sets.kill(&ResActIndex::active(borrow_index));
+                    }
+                }
+            }
+            mir::TerminatorKind::SwitchInt {..} |
+            mir::TerminatorKind::Drop {..} |
+            mir::TerminatorKind::DropAndReplace {..} |
+            mir::TerminatorKind::Call {..} |
+            mir::TerminatorKind::Assert {..} |
+            mir::TerminatorKind::Yield {..} |
+            mir::TerminatorKind::Goto {..} |
+            mir::TerminatorKind::FalseEdges {..} |
+            mir::TerminatorKind::Unreachable => {}
+        }
+
+        // Now we have finished computing the effect of the terminator
+        // on the activations.
+        //
+        // But before we return, we need to establish the terminator's
+        // effect on the reservations. (borrow_check relies on the
+        // reservation state represented alongside the activations in
+        // the ActiveBorrows, even at the end of basic blocks.)
+        self.terminator_effect_on_reservations(sets, location);
+    }
+}
+
+impl<'a, 'gcx, 'tcx> ActiveBorrows<'a, 'gcx, 'tcx> {
+    pub(crate) fn borrows(&self) -> &IndexVec<BorrowIndex, BorrowData<'tcx>> {
+        self.0.borrows()
+    }
+
+    /// Returns the span for the "end point" given region. This will
+    /// return `None` if NLL is enabled, since that concept has no
+    /// meaning there.  Otherwise, return region span if it exists and
+    /// span for end of the function if it doesn't exist.
+    pub(crate) fn opt_region_end_span(&self, region: &Region) -> Option<Span> {
+        match self.0.nonlexical_regioncx {
+            Some(_) => None,
+            None => {
+                match self.0.region_span_map.get(region) {
+                    Some(span) => Some(span.end_point()),
+                    None => Some(self.0.mir.span.end_point())
+                }
+            }
+        }
+    }
+}
+
+struct FindPlaceUses<'a, 'b: 'a, 'tcx: 'a> {
+    assigned_map: &'a FxHashMap<Place<'tcx>, FxHashSet<BorrowIndex>>,
+    sets: &'a mut BlockSets<'b, ResActIndex>,
+}
+
+impl<'a, 'b, 'tcx> FindPlaceUses<'a, 'b, 'tcx> {
+    fn has_been_reserved(&self, b: &BorrowIndex) -> bool {
+        self.sets.on_entry.contains(&ResActIndex::reserved(*b))
+    }
+
+    fn is_potential_use(context: PlaceContext) -> bool {
```

Comment:
Nit: a short comment like 'Given the `PlaceContext` in which a place was found, returns whether it should be considered a "use" of that place (i.e., whether it should activate any associated borrows).' would be nice